### PR TITLE
lib: sms: Add new SMS module

### PR DESCRIPTION
**Pull request by metc**
The SMS module allows to have multiple SMS clients in the system. It
acts as a central SMS client and dispatch SMS data to multiple
listeners. This module acknowledge SMS on the behalf on each listeners.

Multiple listener can register and get notified when a SMS is received.
SMS are read using AT commands, using the shared AT socket.

This module is needed and will be used by the `lwm2m_carrier` library.
I have tested this code using a sample application, but the sample is not yet ready to be added to the -nrf repository.

### CODE REVIEW

File: lib/sms/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+zephyr_include_directories(.)
```
Comment (lemrey) [inline lib/sms/CMakeLists.txt]: This should not be necessary.

File: lib/sms/Kconfig [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+menuconfig SMS
+	bool "SMS subscriber"
+	select AT_CMD
+	select AT_CMD_PARSER
+	select AT_NOTIF
```
Comment (SebastianBoe) [inline lib/sms/Kconfig]: We need some kind of dependency, e.g. on bsd lib or nrf91 to hide this from nrf52 platforms.

Comment (SebastianBoe) [reply lib/sms/Kconfig]: `	depends on TRUSTED_EXECUTION_NONSECURE
`

is what bsd library uses and is as good as any I suppose.

Comment (SebastianBoe) [reply lib/sms/Kconfig]: Ah, no, then it will be visible on nrf53 ...

File: lib/Kconfig [hunk 3]
```diff
@@ -6,22 +6,24 @@
 
 menu "Libraries"
 
-rsource "bin/Kconfig"
+rsource "at_cmd_parser/Kconfig"
```
Comment (lemrey) [inline lib/Kconfig]: Could you split these changes and those to `lib/CMakeLists.txt` into a separate commit?

File: lib/sms/sms.c [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
+	(void)at_params_size_get(&resp_list, 1, &alpha_len);
```
Comment (lemrey) [inline lib/sms/sms.c]: I would only add `(void)` if the compiler is complaining.

File: lib/sms/sms.c [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	if (cmt_rsp.alpha != NULL) {
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
```
Comment (lemrey) [inline lib/sms/sms.c]: I would put declarations at the beginning of a function, it's easier to see how much data is allocated on the stack and to get an idea of what the function will try to do next.

File: lib/sms/sms.c [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+struct sms_subscriber {
+	/* Listener user context. */
+	void *ctx;
+	/* Listener callback. */
+	sms_callback_t listener;
+};
+
+/** @brief List of subscribers. */
+static struct sms_subscriber subsribers[CONFIG_SMS_MAX_SUBSCRIBERS_CNT];
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: subsribers

typo

File: lib/sms/sms.c [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+	cmt_rsp.pdu[pdu_len] = '\0';
+
+	return 0;
+}
+
+/** @brief Handler for AT responses and unsolicited events. */
+void sms_at_handler(void *context, char *at_notif)
+{
+	ARG_UNUSED(context);
+	LOG_DBG("AT notification '%s'", log_strdup(at_notif));
```
Comment (lemrey) [inline lib/sms/sms.c]: I would move this down so that we only log something when it's an SMS,
AT notifications are logged somewhere else already.

File: include/sms.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+struct sms_data;
+
+/** @brief SMS listener callback function. */
+typedef void (*sms_callback_t)(struct sms_data *const data, void *context);
+
+/**
+ * @brief SMS subscriber handle.
+ * A negative value indicates an invalid handle.
+ */
+typedef ssize_t sms_handle_t;
```
Comment (SebastianBoe) [inline include/sms.h]: int<num>_t seems more natural to use here since this is an identifier, and not a size.

File: CODEOWNERS [hunk 9]
```diff
@@ -48,13 +48,14 @@ Kconfig*                                  @SebastianBoe
 /include/event_manager.rst                @b-gent
 /include/profiler.rst                     @b-gent
 /lib/bin/                                 @rlubos @lemrey
-/lib/at_cmd_parser/                       @rlubos
+/lib/at_cmd_parser/                       @rlubos @metc
```
Comment (lemrey) [inline CODEOWNERS]: Please split this into a separate commit.

File: include/sms.h (hunk 8 above)
Comment (lemrey) [reply include/sms.h]: I would avoid the `typedef` and use `int`, you can call the param `sms_handle` in the API.

File: lib/sms/sms.c [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+	if (valid_notif != 0) {
+		LOG_ERR("Invalid SMS notification format");
+		return;
+	}
+
+	/* Notify all subsribers. */
+	LOG_DBG("Valid SMS notification decoded");
+	for (sms_handle_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: subsribers[i]

and change
sms_handle_t i
to
int i
or
uint32_t i

I wouldn't bother making a variable for subsribers[i], so just get rid of the variable s and inline subsribers[i].

File: lib/sms/sms.c [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+
+	ret = at_parser_max_params_from_str(resp, NULL, &resp_list,
+					    AT_CNMI_PARAMS_COUNT);
+	if (ret) {
+		LOG_INF("%s", log_strdup(resp));
+		LOG_ERR("Invalid AT response, err: %d", ret);
+		return ret;
+	}
+
+	int cnmi = 0, value;
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: value has too much scope, put it inside the for loop.

File: lib/sms/sms.c [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+		if (ret) {
+			LOG_ERR("Invalid AT response parameters, err: %d", ret);
+			return ret;
+		}
+		cnmi += value;
+	}
+
+	/* No client are registered when the first 4 parameters are 0. */
+	if (cnmi != 0) {
+		LOG_ERR("Only one SMS client can be registred");
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: registred

typo

File: lib/sms/sms.c [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+	int cnmi = 0, value;
+
+	for (int i = 1; i < AT_CNMI_PARAMS_COUNT - 1; i++) {
+		ret = at_params_int_get(&resp_list, i, &value);
+		if (ret) {
+			LOG_ERR("Invalid AT response parameters, err: %d", ret);
+			return ret;
+		}
+		cnmi += value;
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: Can value be negative?

If not it suffices to

error out here instead of accumulating value and erroring out later.

File: lib/sms/sms.c [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+	}
+
+	/* No client are registered when the first 4 parameters are 0. */
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: The comment says "the first 4", but we are checking 6.

File: lib/sms/sms.c [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+		return ret;
+	}
+
+	/* Clear all observers. */
+	for (size_t i = 0; i < ARRAY_SIZE(subsribers); i++) {
+		const struct sms_subscriber *s = subsribers + i;
+
+		s->ctx = NULL;
+		s->listener = NULL;
+	}
```
Comment (SebastianBoe) [inline lib/sms/sms.c]: Same feedback as here:

https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1633/files#r356565136

interestingly we use a different type for i here.

File: include/sms.h (hunk 8 above)
Comment (metc) [reply include/sms.h]: This handle identifies the (handler, context) pair. This is why I would like to use a new type. Is this allowed?
My implementation uses the index of an array as the handle, this is why I choose to use `size_t`. Any `int_t` size could work as well. It will just limit the number of subscribers.

Comment (rlubos) [reply include/sms.h]: My two cents,

it's not disallowed to add new types, it's discouraged though - that's why you get the following warning: `WARNING:NEW_TYPEDEFS: do not add new typedefs`.

My personal opinion is that covering simple types with custom typedefs brings more bad than good - it does not provide any extra type protection (you can mixup `sms_handle_t`/`int`/`ssize_t` however you want, compiler won't complain), yet it makes the code harder to analyze (as you always need to check what exactly the type is whenever you encounter a new type). So in this case, I'd back @lemrey opinion.

File: lib/sms/sms.c [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+
+		if (s->ctx == NULL && s->listener == NULL) {
+			s->ctx = context;
+			s->listener = listener;
+			return i;
+		}
+	}
+
+	/* Too many subsribers. */
+	return -1;
```
Comment (rlubos) [inline lib/sms/sms.c]: Shall we return the actual error code here? The function itself is pretty straightforward, still it would be good to differentiate between `-EINVAL`/`-ENOMEM` here.

**Comment by rlubos** [comment]
BTW, will any sample or tests using this module follow? For now the CI won't even check if this code compiles.

File: include/sms.h (hunk 8 above)
Comment (SebastianBoe) [reply include/sms.h]: > My implementation uses the index of an array as the handle, this is why I choose to use size_t. Any int_t size could work as well. It will just limit the number of subscribers.

My understanding is that size_t should be used for sizes of things, an index into an array is not a size and should therefore not be used. It's not clear to me how this will limit the number of subscribers.

Comment (SebastianBoe) [reply include/sms.h]: My understanding was wrong, using size_t for indexes is fine.

File: include/sms.h [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+ * @return Zero on success, or a negative error code. The EBUSY error
+ *         indicates that one SMS client has already been registered.
+ */
+int sms_init(void);
+
+/**
+ * @brief Register a new listener.
+ *
+ * A listener is identified by a unique handle value. This handle should be used
+ * to unregister the listener. A listener can be registered multiple times with
```
[the remark this answers is not in the document]
Comment (metc) [reply include/sms.h]: I think it is a limitation if we only use the function pointer to unregister.
Each subscriber is identified by a context and a function callback, meaning the same module can subscribe several times using a different context (with the same or a different callback function). I think this is a valid use case.

The other option is to always give the function callback and the context (two parameters) to unregister has is it done in https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/include/at_notif.h#L79. In this case, we do not need to save the handle, but the context.
The bsdlib uses handles, so I choose to use the same approach. In simple applications, we do not need to unregister at runtime, so we can just ignore the handle value.

**Comment by metc** [comment]
> BTW, will any sample or tests using this module follow? For now the CI won't even check if this code compiles.

> This module is needed and will be used by the lwm2m_carrier library.
I have tested this code using a sample application, but the sample is not yet ready to be added to the -nrf repository.

The sample application is not ready yet. This module will be used by the lwm2m_carrier lib to allow our customers to send/received SMS when LWM2M is used.

File: CODEOWNERS (hunk 9 above)
Comment (metc) [reply CODEOWNERS]: Do I need a new commit because of the changes in line 51 or because of line 58?
In this PR, all the changes have been done in one commit: https://github.com/NordicPlayground/fw-nrfconnect-nrf/commit/67ded39d123cf6b3b7ae14150d180842bf0cb721#diff-5daf978e60e03b3975b485d5f1b449dc

File: include/sms.h (hunk 8 above)
Comment (metc) [reply include/sms.h]: I like to use `typedef` to keep the API binary compatible.
I will use and `int` as we do in the `bsdlib` to represent handles.

File: lib/Kconfig (hunk 3 above)
Comment (metc) [reply lib/Kconfig]: I think it is easier to have this list sorted. I will to this in a different commit.

File: lib/sms/sms.c (hunk 7 above)
Comment (metc) [reply lib/sms/sms.c]: Yes, it can be enabled from the `at_cmd` module (https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/7e6228bb7e451ff19628636506b7caf3995ca1ec/drivers/at_cmd/at_cmd.c#L174).
I will remove lines 149 and 153 then.

File: lib/sms/sms.c (hunk 10 above)
Comment (metc) [reply lib/sms/sms.c]: The compiler will inline it for me, but I will change it.

**Comment by ru-fu** [comment]
This looks like it needs documentation.

File: lib/sms/Kconfig (hunk 2 above)
Comment (metc) [reply lib/sms/Kconfig]: Actually, the KConfig dependencies for `bsdlib` are wrong.
Using `SOC_SERIES_NRF91X` has a dependency to `BOARD_NRF9160_PCA10090` as well (secure).
I do not think I can use `BOARD_NRF9160_PCA10090` because this module and `bsdlib` do run on the Thingy:91 as well. Any thoughts?

File: lib/sms/sms.c (hunk 14 above)
Comment (metc) [reply lib/sms/sms.c]: I am reading the parameter indexes [1..4].

File: lib/sms/sms.c (hunk 13 above)
Comment (metc) [reply lib/sms/sms.c]: I will simplify the logic as you proposed, thank you.

File: lib/sms/sms.c (hunk 12 above)
Comment (metc) [reply lib/sms/sms.c]: Typos fixed, and spell checker as well :)

File: lib/sms/sms.c (hunk 16 above)
Comment (metc) [reply lib/sms/sms.c]: Good point. Fixed.

File: lib/sms/sms.c (hunk 11 above)
Comment (metc) [reply lib/sms/sms.c]: Fixed

**Comment by metc** [comment]
> This looks like it needs documentation.

Module documentation added. Please have a look.

**Comment by ru-fu** [comment]
I added my updates - please check and squash, if okay.
The doc CI will get red for this PR. Please merge in the current master; this should fix the doc errors.

File: include/sms.h [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ * the same or a different context.
+ *
+ * @param listener Callback function. Cannot be null.
+ * @param context User context. Can be null if not used.
+ *
+ * @retval -EINVAL Invalid parameter.
+ * @retval -ENOMEM No memory to register new observers.
+ * @return Handle identifying the listener,
```
[the remark this answers is not in the document]
Comment (metc) [reply include/sms.h]: I would like to list the error codes here. How should I do this then?

File: include/sms.h [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <sys/types.h>
+
+/* Forward declaration. */
```
[the remark this answers is not in the document]
Comment (metc) [reply include/sms.h]: Of course. My mistake, this is an incomplete type. The sample application leaves in a different branch, I forget to push this change. Fixed now, thank you.

[the remark this answers is not in the document]
Comment (metc) [reply include/sms.h]: Of course. My mistake, this is an incomplete type. The sample application leaves in a different branch, I forget to push this change. Fixed now, thank you.
